supervisor/homeassistant/module.py

```diff
@@ -313,46 +322,114 @@ async def backup(self, secure_tar_file: SecureTarFile) -> None:
-                )
+                write_json_file(temp_path.joinpath("homeassistant.json"), self._data)
+            except ConfigurationFileError as err:
+                raise HomeAssistantError(
+                    f"Can't save meta for Home Assistant Core: {err!s}", _LOGGER.error
+                ) from err
+
+            # Backup data config folder
+            def _write_tarfile():
+                with tar_file as backup:
+                    # Backup metadata
+                    backup.add(temp, arcname=".")
+
+                    # Backup data
+                    atomic_contents_add(
+                        backup,
+                        self.sys_config.path_homeassistant,
+                        excludes=HOMEASSISTANT_BACKUP_EXCLUDE,
+                        arcname="data",
+                    )
 
-    async def restore(self, secure_tar_file: SecureTarFile) -> None:
+            try:
+                _LOGGER.info("Backing up Home Assistant Core config folder")
+                await self.sys_run_in_executor(_write_tarfile)
+                _LOGGER.info("Backup Home Assistant Core config folder done")
+            finally:
+                try:
+                    await self.sys_homeassistant.websocket.async_send_command(
+                        {ATTR_TYPE: WSType.BACKUP_END}
+                    )
+                except HomeAssistantWSError:
+                    _LOGGER.warning(
+                        "Error during Home Assistant Core backup. Check HA Core logs."
+                    )
+
+    async def restore(self, tar_file: tarfile.TarFile) -> None:
         """Restore Home Assistant Core config/ directory."""
+        with TemporaryDirectory(dir=self.sys_config.path_tmp) as temp:
+            temp_path = Path(temp)
+            temp_data = temp_path.joinpath("data")
+            temp_meta = temp_path.joinpath("homeassistant.json")
 
-        # Perform a restore
-        def _restore_tarfile():
-            origin_dir = Path(self.sys_config.path_supervisor, FOLDER_HOMEASSISTANT)
+            # extract backup
+            def _extract_tarfile():
+                """Extract tar backup."""
+                with tar_file as backup:
+                    backup.extractall(path=temp_path, members=secure_path(backup))
 
-            with secure_tar_file as tar_file:
-                tar_file.extractall(path=origin_dir, members=tar_file)
+            try:
+                await self.sys_run_in_executor(_extract_tarfile)
+            except tarfile.TarError as err:
+                raise HomeAssistantError(
+                    f"Can't read tarfile {tar_file}: {err}", _LOGGER.error
+                ) from err
+
+            # Check old backup format v1
+            if not temp_data.exists():
+                temp_data = temp_path
+
+            # Restore data
+            def _restore_data():
+                """Restore data."""
+                shutil.copytree(
```

Comment:
This throws when I test it
```
22-02-23 15:25:30 ERROR (MainThread) [supervisor.backups.manager] Restore 1d4f9d05 error
Traceback (most recent call last):
  File "/usr/src/supervisor/supervisor/backups/manager.py", line 254, in _do_restore
    task_hass = await backup.restore_homeassistant()
  File "/usr/src/supervisor/supervisor/backups/backup.py", line 478, in restore_homeassistant
    await self.sys_homeassistant.restore(homeassistant_file)
  File "/usr/src/supervisor/supervisor/homeassistant/module.py", line 398, in restore
    await self.sys_run_in_executor(_restore_data)
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/src/supervisor/supervisor/homeassistant/module.py", line 391, in _restore_data
    shutil.copytree(
  File "/usr/local/lib/python3.9/shutil.py", line 566, in copytree
    return _copytree(entries=entries, src=src, dst=dst, symlinks=symlinks,
  File "/usr/local/lib/python3.9/shutil.py", line 467, in _copytree
    os.makedirs(dst, exist_ok=dirs_exist_ok)
  File "/usr/local/lib/python3.9/os.py", line 225, in makedirs
    mkdir(name, mode)
FileExistsError: [Errno 17] File exists: '/data/homeassistant'
```

Reply:
Hehe, I actually run into this the very same way! I think `stress-ng` creates immutable files or something.

Reply:
Yup :/ I need to deal with that one day...